backend/cfg/cfg_regalloc_utils.ml

```diff
@@ -450,3 +455,83 @@ let may_use_stack_operands_everywhere :
   may_use_stack_operands_array map instr.res;
   All_spilled_registers_rewritten
+
+let insert_block :
+    Cfg_with_layout.t ->
+    Cfg.BasicInstructionList.t ->
+    after:Cfg.basic_block ->
+    next_instruction_id:(unit -> Instruction.id) ->
+    unit =
+ fun cfg_with_layout body ~after:predecessor_block ~next_instruction_id ->
+  let cfg = Cfg_with_layout.cfg cfg_with_layout in
+  let successors =
+    Cfg.successor_labels ~normal:true ~exn:false predecessor_block
+  in
+  (* CR gyorsh: this function is currently only when |successors|>=1 and |body|
+     > 0*)
+  if Label.Set.cardinal successors = 0
+  then
+    Misc.fatal_errorf
+      "Cannot insert a block after block %a: it has no successors" Label.print
+      predecessor_block.start;
+  let last_insn =
+    match Cfg.BasicInstructionList.last body with
+    | None -> Misc.fatal_error "Inserting an empty block"
+    | Some i -> i
+  in
+  let copy (i : Cfg.basic Cfg.instruction) : Cfg.basic Cfg.instruction =
+    { i with id = next_instruction_id () }
+  in
+  (* copy body if there is more than one successor *)
+  let first = ref true in
+  let get_body () =
+    if !first
+    then (
+      first := false;
+      body)
+    else
+      let new_body = Cfg.BasicInstructionList.make_empty () in
+      Cfg.BasicInstructionList.iter body ~f:(fun instr ->
+          Cfg.BasicInstructionList.add_end new_body (copy instr));
+      new_body
+  in
+  Label.Set.iter
+    (fun successor_label ->
+      let successor_block = Cfg.get_block_exn cfg successor_label in
+      let start = Cmm.new_label () in
+      let block : Cfg.basic_block =
+        { start;
+          body = get_body ();
+          terminator =
+            { (* The [successor_block] is the only successor. *)
+              desc = Cfg.Always successor_label;
+              arg = [||];
+              res = [||];
+              dbg = last_insn.dbg;
+              fdo = last_insn.fdo;
+              live = last_insn.live;
+              stack_offset = last_insn.stack_offset;
+              id = next_instruction_id ();
+              irc_work_list = Unknown_list
+            };
+          (* The [predecessor_block] is the only predecessor. *)
+          predecessors = Label.Set.singleton predecessor_block.start;
+          stack_offset = predecessor_block.terminator.stack_offset;
+          exn = None;
+          can_raise = false;
+          is_trap_handler = false;
+          dead = predecessor_block.dead
+        }
+      in
+      Cfg_with_layout.add_block cfg_with_layout block
+        ~after:predecessor_block.start;
+      (* Change the labels for the terminator in [predecessor_block]. *)
+      Cfg.replace_successor_labels cfg ~normal:true ~exn:false predecessor_block
+        ~f:(fun old_label ->
+          if Label.equal old_label successor_label then start else old_label);
+      (* Update predecessors for the [successor_block]. *)
+      successor_block.predecessors
+        <- successor_block.predecessors
```

Comment:
I am slightly confused, but isn't it
`singleton start`?

Reply:
`block.predecessors` is a singleton, but `successor_block.predecessors` can be anything.

Reply:
Especially if something different that predecesoor_block jumped there before insertion it will also jump after.

Reply:
I will take your word for it.

Reply:
If `successor_block.precesessor == Set.from_list [predecessor_block]` is not a requirement overall, it's a condition for optimization. Then we don't have to add new block, we can simply add all the spills to the `successor_block` directly.